planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+  smaller tables to work through
+
+### Projections
+
+Thoughts:
+
+Functionality wise, the lifecycle of projections are:
+
+* "Live" projections built on demand from the raw events. This only changes for v4 to take advantage of new snapshotting capabilities
+* "Inline" projections built as part of the same unit of work as capturing the source events. See the section on *Inline Projections* because this
+  will change internally for v4
+* "Async" projections built up in a background process of some sort. I think this turns into a near re-write for v4
+
+By *type*, the projection types *could* be:
+
+* Aggregate document by stream (existing)
+* Aggregate document by tenant id (new)
+* Execute SQL statement by event type (new) -- so, insert/update/delete. This comes up a lot, and it should be low hanging fruit
+* Transform a single event to a document --> needs to become just "carry out some kind of change to a document" and probably ends up sharing more mechanics with the aggregation than it used to
+* Aggregate by arbitrary values of either the stream or event types -- but if we do this, I strongly vote to persist whatever the differentiation value is in the database to make projection rebuilds and the async daemon more "parallelizable"
+
+
+Other thoughts:
+
+* I want to completely throw away the current `Aggregator` implementations. It was just too naive for real world usage and the internals have gotten ugly
+* Re-evaluate the `ViewProjection` fluent interface. That might fit in fine to our new model of projections for v4 with more overloads
+* I've been in favor of having adapters for existing projection libraries like [Liquid Projections](https://liquidprojections.net/) in the past, but I think we'd be losing too
+  much optization, especially for the aggregations inside the async daemon.
+
+#### Defining a Projection to a Document
+
+
+```
+ProjectTo<ProjectedDocument>()
+    // How to transform the events to the projected document, see below
+    .TransformWith<SampleTransformer>()
+
+    // Optionally aggregate by stream
+    .AggregateByStream() // Think this is self-descriptive
+
+    .AggregateByTenant() // Not sure how common this would be. Aggregate across the entire tenant
+
+    // Or aggregate by something special, hopefull use a common base type here for the event type
+    .AggregateByEventData<TEvent>(Expression<Func<TEvent, int>> expression) // what's the identity of the projected, aggregate doc?
+
+    .DeleteOn<TEvent>() // the aggregate document should be deleted if an event of this type hits
+                        // In this case, we could look ahead and "know" in the rebuilds not to rebuild
+                        // a projection for this aggregate as an optimization
+
+    .SnapshotInIncrementsOf(10) // Then maybe there are other options to retain old snapshots or not? Dunno
```

Comment:
I think that we should support much more flexible options for snapshots. Once per number of events might be in theory useful for performance, but I'm not sure if it'll be used as it's hard to predict the corellation to the business actions (so if it should be snapshoted or not). The more functional would be imho snapshot on exact event type. See eg. https://verraes.net/2019/05/patterns-for-decoupling-distsys-passage-of-time-event/